### PR TITLE
#8251 DisplaySnackbar

### LINKED ISSUE #8251: [Enhancement] Application.Current.MainPage.DisplaySnackbar
## Summary
MainPage can display messages and prompts, which requires user interaction (press button).
Using toast or snackbar you can display user friendly notification without necessity to click any button. 

```Application.Current.MainPage.DisplaySnackbar(message, duration = 3000, actionButtonText=null, action=null);```

## API Changes

Add new method which will call native implementation:
- For Android I think it would be Snackbar (https://material.io/develop/android/components/snackbar/) or Toast (https://developer.android.com/guide/topics/ui/notifiers/toasts)
- For iOS something like this (https://www.cocoacontrols.com/controls/snackbar-ios)
- For UWP/macOS I vote for Notifications as a native way to display any info (but snackbar would also look very nice)
- WPF/GTK/Tizen - up to devs - I think snackbar would be easier implement

## Intended Use Case
Any kind of notifications ("Data is stored; File is deleted");

### PR DESCRIPTION
**Pull request by VladislavAntonyuk**
### Issues Resolved ### 

- fixes #8251

### API Changes ###

Added:
 - public Task<bool> DisplaySnackbar(string message, int duration = 3000, string actionButtonText = null, Func<Task> action = null)
- Implementation for each platform (Snackbar for Droid, UIView for iOS, Windows Notification for UWP, MsgBox with autoclose for GTK, GridRow for WPF, Native.Dialog Tizen)

### Platforms Affected ### 

- All platforms

### Before/After Screenshots ### 
- UWP
image

- Tizen
image

- macOS
image

- GTK
image

- iOS
image

- WPF
image

### Testing Procedure ###
Issue8251 in ControlGallery

### PR Checklist ###

- [x] Targets the correct branch
- [x] Tests are passing (or failures are unrelated)

- [x] Android
- [x] Tizen
- [x] iOS
- [x] macOS
- [x] UWP
- [x] WPF
- [x] GTK

### CODE REVIEW

File: Xamarin.Forms.ControlGallery.iOS/Xamarin.Forms.ControlGallery.iOS.csproj [hunk 1]
```diff
@@ -22,7 +22,7 @@
     <WarningLevel>4</WarningLevel>
     <ConsolePause>false</ConsolePause>
     <MtouchArch>i386, x86_64</MtouchArch>
-    <MtouchLink>SdkOnly</MtouchLink>
+    <MtouchLink>None</MtouchLink>
```
Comment (rmarinho) [inline Xamarin.Forms.ControlGallery.iOS/Xamarin.Forms.ControlGallery.iOS.csproj]: We should revert this change

File: Xamarin.Forms.ControlGallery.Tizen/Xamarin.Forms.ControlGallery.Tizen.csproj [hunk 2]
```diff
@@ -1,18 +1,13 @@
-﻿<Project Sdk="Microsoft.NET.Sdk">
+﻿<Project Sdk="Tizen.NET.Sdk/1.0.8">
 
   <PropertyGroup>
     <OutputType>Exe</OutputType>
-    <TargetFramework>tizen40</TargetFramework>
+    <TargetFramework>tizen50</TargetFramework>
```
Comment (rmarinho) [inline Xamarin.Forms.ControlGallery.Tizen/Xamarin.Forms.ControlGallery.Tizen.csproj]: Not sure we want to update this, do we need to update tizen ?

File: Xamarin.Forms.Core/ToastArguments.cs [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+		public ToastArguments(string message, int duration, string actionButtonText, Func<Task> action)
+		{
+			Duration = duration;
+			Message = message;
+			ActionButtonText = actionButtonText;
+			Action = action;
+			Result = new TaskCompletionSource<bool>(false);
+		}
+
+		/// <summary>
```
Comment (rmarinho) [inline Xamarin.Forms.Core/ToastArguments.cs]: We can drop these inline comments

File: Xamarin.Forms.Platform.iOS/Controls/Snackbar/Extensions/UIViewExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+
+namespace Xamarin.Forms.Platform.iOS.Controls.Snackbar.Extensions
+{
+	/// <summary>
+	/// Provides pre iOS 11 compatible helper properties for for SafeAreaLayoutGuide anchors.
+	/// Helper properties has prefix 'Safe' in their name. 
+	/// If app is running on pre iOS 11, Anchor properties from the view will be used.
+	/// If app is running on iOS 11 or grater, SafeAreaLayoutGuide properties will be used.
+	/// </summary>
+	public static class UIViewExtensions
```
Comment (rmarinho) [inline Xamarin.Forms.Platform.iOS/Controls/Snackbar/Extensions/UIViewExtensions.cs]: The code style is wrong, we use tabs, you can also remove the summary.

I think we should do this internal.

File: Xamarin.Forms.Platform.Tizen/PopupManager.cs [hunk 5]
```diff
@@ -88,6 +91,58 @@ void OnBusySetRequest(Page sender, bool enabled)
 			}
 		}
 
+		Timer toastTimer;
+		Native.Dialog toastDialog;
+		void OnToastRequest(Page sender, ToastArguments arguments)
```
Comment (rmarinho) [inline Xamarin.Forms.Platform.Tizen/PopupManager.cs]: Opening 2 times on Tizen isn't working for me.

File: Xamarin.Forms.Platform.UAP/Platform.cs [hunk 6]
```diff
@@ -46,7 +48,7 @@ public static IVisualElementRenderer CreateRenderer(VisualElement element)
 				throw new ArgumentNullException(nameof(element));
 
 			IVisualElementRenderer renderer = Registrar.Registered.GetHandlerForObject<IVisualElementRenderer>(element) ??
-			                                  new DefaultRenderer();
+											  new DefaultRenderer();
```
Comment (rmarinho) [inline Xamarin.Forms.Platform.UAP/Platform.cs]: Not sure about these style changes

File: Xamarin.Forms.Platform.UAP/Platform.cs [hunk 7]
```diff
@@ -297,16 +299,16 @@ async void SetCurrent(Page newPage, bool popping = false, Action completedCallba
...
+				// library might still be retrieving it and using it
+				newPage.Platform = this;
 #pragma warning restore CS0618 // Type or member is obsolete
 
-			if (_currentPage != null)
-			{
-				Page previousPage = _currentPage;
-				IVisualElementRenderer previousRenderer = GetRenderer(previousPage);
-				_container.Children.Remove(previousRenderer.ContainerElement);
+				if (_currentPage != null)
```
Comment (rmarinho) [inline Xamarin.Forms.Platform.UAP/Platform.cs]: are these style changes needed ?

File: Xamarin.Forms.Platform.UAP/Platform.cs [hunk 8]
```diff
@@ -492,11 +494,59 @@ internal IToolbarProvider GetToolbarProvider()
 
 		internal static void SubscribeAlertsAndActionSheets()
 		{
+			MessagingCenter.Subscribe<Page, ToastArguments>(Window.Current, Page.ToastSignalName, OnPageToast);
 			MessagingCenter.Subscribe<Page, AlertArguments>(Window.Current, Page.AlertSignalName, OnPageAlert);
 			MessagingCenter.Subscribe<Page, ActionSheetArguments>(Window.Current, Page.ActionSheetSignalName, OnPageActionSheet);
 			MessagingCenter.Subscribe<Page, PromptArguments>(Window.Current, Page.PromptSignalName, OnPagePrompt);
 		}
 
+		private static void OnPageToast(Page page, ToastArguments args)
```
Comment (rmarinho) [inline Xamarin.Forms.Platform.UAP/Platform.cs]: We should drop private

File: Xamarin.Forms.Platform.UAP/Xamarin.Forms.Platform.UAP.csproj [hunk 9]
```diff
@@ -101,13 +101,16 @@
       <Generator>MSBuild:Compile</Generator>
     </Page>
   </ItemGroup>
+  <ItemGroup>
```
Comment (rmarinho) [inline Xamarin.Forms.Platform.UAP/Xamarin.Forms.Platform.UAP.csproj]: Can we do this without adding this dependency ?

File: Xamarin.Forms.Platform.WPF/Platform.cs [hunk 10]
```diff
@@ -4,21 +4,24 @@
...
 {
-	public class Platform : BindableObject, INavigation
+	public class Platform : BindableObject, INavigation, IDisposable
 #pragma warning disable CS0618
 		, IPlatform
 #pragma warning restore
 	{
 		readonly FormsApplicationPage _page;
+
+		private bool _disposed;
```
Comment (rmarinho) [inline Xamarin.Forms.Platform.WPF/Platform.cs]: drop private

File: Xamarin.Forms.Platform.WPF/Controls/FormWindow.cs [hunk 11]
```diff
@@ -35,6 +36,26 @@ public class FormsWindow : Window
 		public static readonly DependencyProperty CurrentContentDialogProperty = DependencyProperty.Register("CurrentContentDialog", typeof(FormsContentDialog), typeof(FormsWindow));
 		public static readonly DependencyProperty TitleBarBackgroundColorProperty = DependencyProperty.Register("TitleBarBackgroundColor", typeof(Brush), typeof(FormsWindow));
 		public static readonly DependencyProperty TitleBarTextColorProperty = DependencyProperty.Register("TitleBarTextColor", typeof(Brush), typeof(FormsWindow));
+		public static readonly DependencyProperty SnackbarActionCommandProperty = DependencyProperty.Register("SnackbarActionCommand", typeof(ICommand), typeof(FormsWindow));
```
Comment (rmarinho) [inline Xamarin.Forms.Platform.WPF/Controls/FormWindow.cs]: Why do we need these public properties ?

File: Xamarin.Forms.Platform.iOS/Controls/Snackbar/SnackBar.cs [hunk 12]
```diff
@@ -0,0 +1,120 @@
+﻿using System;
+using Xamarin.Forms.Platform.iOS.Controls.Snackbar.SnackbarViews;
+using Foundation;
+using System.Threading.Tasks;
+using UIKit;
+
+namespace Xamarin.Forms.Platform.iOS.Controls.Snackbar
+{
+	public class SnackBar
```
Comment (rmarinho) [inline Xamarin.Forms.Platform.iOS/Controls/Snackbar/SnackBar.cs]: Does it make sense to call it SnackBar? or should we call it Toast?

File: Xamarin.Forms.Platform.iOS/Controls/Snackbar/SnackBar.cs [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+	{
+		public string Message { get; protected set; }
+		public double Duration { get; protected set; }
+		public string ActionButtonText { get; protected set; }
+
+		private NSTimer timer;
```
Comment (rmarinho) [inline Xamarin.Forms.Platform.iOS/Controls/Snackbar/SnackBar.cs]: drop private

File: Xamarin.Forms.Platform.iOS/Controls/Snackbar/SnackbarViews/ActionMessageSnackbarView.cs [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+{
+    public class ActionMessageSnackbarView : MessageSnackbarView
+    {
+		readonly SnackBar _snackbar;
+
+		public UIButton ActionButton { get; set; }
+
+        public ActionMessageSnackbarView(SnackBar snackbar) : base(snackbar)
+        {
+			_snackbar = snackbar;
```
Comment (rmarinho) [inline Xamarin.Forms.Platform.iOS/Controls/Snackbar/SnackbarViews/ActionMessageSnackbarView.cs]: Style is wrong,

File: Xamarin.Forms.Platform.iOS/Controls/Snackbar/SnackbarViews/ActionMessageSnackbarView.cs [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+
+            ActionButton = new UIButton(UIButtonType.System);
+            ActionButton.SetTitle(Snackbar.ActionButtonText, UIControlState.Normal);
+            ActionButton.TranslatesAutoresizingMaskIntoConstraints = false;
+            ActionButton.TitleLabel.LineBreakMode = Snackbar.Appearance.DismissButtonLineBreakMode;
+            ActionButton.TouchUpInside += DismissButton_TouchUpInside;
+            AddSubview(ActionButton);
+        }
+
+        /// <summary>
```
Comment (rmarinho) [inline Xamarin.Forms.Platform.iOS/Controls/Snackbar/SnackbarViews/ActionMessageSnackbarView.cs]: We don't use the iniline comments

File: Xamarin.Forms.Platform.iOS/Controls/Snackbar/SnackbarViews/BaseSnackbarView.cs [hunk 16]
```diff
@@ -0,0 +1,130 @@
+﻿using UIKit;
+using Xamarin.Forms.Platform.iOS.Controls.Snackbar.Extensions;
+
+namespace Xamarin.Forms.Platform.iOS.Controls.Snackbar.SnackbarViews
+{
+	public class BaseSnackbarView : UIView
```
Comment (rmarinho) [inline Xamarin.Forms.Platform.iOS/Controls/Snackbar/SnackbarViews/BaseSnackbarView.cs]: The code style is wrong, we use tabs, you can also remove the summary.

I think we should do this internal.

File: Xamarin.Forms.Core/ToastArguments.cs [hunk 17]
```diff
@@ -0,0 +1,49 @@
+using System;
+using System.ComponentModel;
+using System.Threading.Tasks;
+
+namespace Xamarin.Forms.Internals
+{
+	[EditorBrowsable(EditorBrowsableState.Never)]
```
Comment (rmarinho) [inline Xamarin.Forms.Core/ToastArguments.cs]: Why do you hide this?

File: Xamarin.Forms.ControlGallery.iOS/Xamarin.Forms.ControlGallery.iOS.csproj (hunk 1 above)
Comment (VladislavAntonyuk) [reply Xamarin.Forms.ControlGallery.iOS/Xamarin.Forms.ControlGallery.iOS.csproj]: VS shows a warning that Hot reload will not work. Why do we use SdkOnly for debug?

File: Xamarin.Forms.Platform.UAP/Platform.cs (hunk 6 above)
Comment (VladislavAntonyuk) [reply Xamarin.Forms.Platform.UAP/Platform.cs]: Ctrl + K 
Ctrl + D

I will review this changes

File: Xamarin.Forms.Platform.UAP/Xamarin.Forms.Platform.UAP.csproj (hunk 9 above)
Comment (VladislavAntonyuk) [reply Xamarin.Forms.Platform.UAP/Xamarin.Forms.Platform.UAP.csproj]: According to Microsoft docs https://docs.microsoft.com/en-us/windows/uwp/design/shell/tiles-and-notifications/send-local-toast

Comment (VladislavAntonyuk) [reply Xamarin.Forms.Platform.UAP/Xamarin.Forms.Platform.UAP.csproj]: I am not sure that notification is a right way to use for DisplayToast. Probably I should implement it the same way as for WPF

File: Xamarin.Forms.Platform.WPF/Controls/FormWindow.cs (hunk 11 above)
Comment (VladislavAntonyuk) [reply Xamarin.Forms.Platform.WPF/Controls/FormWindow.cs]: They are used in FormWindow.xaml

File: Xamarin.Forms.Platform.iOS/Controls/Snackbar/SnackBar.cs (hunk 12 above)
Comment (VladislavAntonyuk) [reply Xamarin.Forms.Platform.iOS/Controls/Snackbar/SnackBar.cs]: I am going to rename the method to DisplaySnackbar. Snack bar = Toast + action button. I hide action button if action is null or action button text is empty

File: Xamarin.Forms.Core/ToastArguments.cs (hunk 17 above)
Comment (VladislavAntonyuk) [reply Xamarin.Forms.Core/ToastArguments.cs]: To be consistent. We use the same attribute for AlertArguments

File: Xamarin.Forms.ControlGallery.Tizen/Xamarin.Forms.ControlGallery.Tizen.csproj (hunk 2 above)
Comment (VladislavAntonyuk) [reply Xamarin.Forms.ControlGallery.Tizen/Xamarin.Forms.ControlGallery.Tizen.csproj]: I was not able to run project with tizen40. also as tizen5 exist and we support it, I decided to update the csproj.

File: Xamarin.Forms.Core/ToastArguments.cs (hunk 3 above)
Comment (VladislavAntonyuk) [reply Xamarin.Forms.Core/ToastArguments.cs]: just want to be consistent with similar file (AlertArguments as example)

File: Xamarin.Forms.Platform.iOS/Controls/Snackbar/SnackbarViews/ActionMessageSnackbarView.cs (hunk 14 above)
Comment (VladislavAntonyuk) [reply Xamarin.Forms.Platform.iOS/Controls/Snackbar/SnackbarViews/ActionMessageSnackbarView.cs]: should I change it to this.snackbar = snackbar?

**Comment by VladislavAntonyuk** [comment]
@rmarinho please let me know what should I add to implement the material snack bar?
I added MaterialSnackbar with ShowSnackbar(SnackbarArguments args) method. 
Snack bar is not a VisualElement so there is not renderer for it.

File: Xamarin.Forms.Platform.UAP/Platform.cs (hunk 6 above)
Comment (VladislavAntonyuk) [reply Xamarin.Forms.Platform.UAP/Platform.cs]: There was a spaces, I replaced them wit tab

File: Xamarin.Forms.ControlGallery.Tizen/Xamarin.Forms.ControlGallery.Tizen.csproj (hunk 2 above)
Comment (rmarinho) [reply Xamarin.Forms.ControlGallery.Tizen/Xamarin.Forms.ControlGallery.Tizen.csproj]: @rookiejava what's your opinion ?

File: Xamarin.Forms.Platform.UAP/Xamarin.Forms.Platform.UAP.csproj (hunk 9 above)
Comment (rmarinho) [reply Xamarin.Forms.Platform.UAP/Xamarin.Forms.Platform.UAP.csproj]: I actually like the native notification, what do you think @samhouts  ?

File: Xamarin.Forms.Platform.WPF/Platform.cs [hunk 18]
```diff
@@ -66,6 +70,31 @@ async void OnPageAlert(Page sender, AlertArguments options)
 			options.SetResult(dialogResult == LightContentDialogResult.Primary);
 		}
 
+		Timer snackbarTimer;
```
Comment (rmarinho) [inline Xamarin.Forms.Platform.WPF/Platform.cs]: _snackbarTimer

File: Xamarin.Forms.Platform.UAP/Xamarin.Forms.Platform.UAP.csproj (hunk 9 above)
Comment (VladislavAntonyuk) [reply Xamarin.Forms.Platform.UAP/Xamarin.Forms.Platform.UAP.csproj]: I am worried about the name of the method. if we decide to add ShowNotification method, it will be the same implementation, so the user won't know is it a notification or a snackbar. But I like Windows Notifications, as it is a native way to show notification, and it is customizable and has action buttons

File: Xamarin.Forms.Platform.UAP/Platform.cs (hunk 7 above)
Comment (VladislavAntonyuk) [reply Xamarin.Forms.Platform.UAP/Platform.cs]: there was invalid indention (3 tabs replaced with 4 tabs). Now all items in the Try block

File: Xamarin.Forms.ControlGallery.Tizen/Xamarin.Forms.ControlGallery.Tizen.csproj (hunk 2 above)
Comment (rmarinho) [reply Xamarin.Forms.ControlGallery.Tizen/Xamarin.Forms.ControlGallery.Tizen.csproj]: I think we need to revert this

File: Xamarin.Forms.ControlGallery.iOS/Xamarin.Forms.ControlGallery.iOS.csproj (hunk 1 above)
Comment (rmarinho) [reply Xamarin.Forms.ControlGallery.iOS/Xamarin.Forms.ControlGallery.iOS.csproj]: Better to make sure sdk linker works when we run tests. Please revert

**Comment by VladislavAntonyuk** [comment]
@rmarinho what do you think about splitting this into 2 features. The first feature is snackbar, the second one is notification. Is notification should be a part of XF or it should be as a separate package (probably Xamarin Essentials)

**Comment by rmarinho** [comment]
@VladislavAntonyuk that would be great. I think going first is put this snackbar in all loking the same.

**Comment by YZahringer** [comment]
Is there a possibility to customize the style of the snackbar?

It would be interesting to be able to customize the style to differentiate information, warning or error messages for example. Maybe just the background color?

It would also be interesting to be able to define the horizontal and vertical position of snackbar.

For example, the Snackbar configuration of the Angular Material component: [MatSnackBarConfig](https://material.angular.io/components/snack-bar/api#MatSnackBarConfig)

**Comment by VladislavAntonyuk** [comment]
@YZahringer, for now I would like to focus on main idea - just display snack bar. Then we can add customization to it

**Comment by VladislavAntonyuk** [comment]
@jsuarezruiz I found PR in Xamarin Essentials related to Notifications. I will revert that changes. As for snackbar I will try to implement for GTK, macOS and UWP. There is no native platform support if it, I will investigate how to inject stacklayout of snackbar to the page.

File: Xamarin.Forms.Platform.iOS/Controls/Snackbar/SnackbarAppearance.cs [hunk 19]
```diff
@@ -0,0 +1,16 @@
+﻿using System;
+using UIKit;
+
+namespace Xamarin.Forms.Platform.iOS.Controls.Snackbar
+{
+	class SnackbarAppearance
+	{
+		public UIColor Color { get; set; } = UIColor.SystemGrayColor;
```
[the remark this answers is not in the document]
Comment (VladislavAntonyuk) [reply Xamarin.Forms.Platform.iOS/Controls/Snackbar/SnackbarAppearance.cs:8]: for now I would like to focus on functionality. We can add customization later

File: Xamarin.Forms.Platform.GTK/Helpers/DialogHelper.cs [hunk 20]
```diff
@@ -34,6 +35,92 @@ public static void ShowAlert(PlatformRenderer platformRender, AlertArguments arg
 			messageDialog.Destroy();
 		}
 
+		static Timer _snackbarTimer;
+		public static void ShowSnackbar(PlatformRenderer platformRender, SnackbarArguments arguments)
+		{
```
[the remark this answers is not in the document]
Comment (VladislavAntonyuk) [reply Xamarin.Forms.Platform.GTK/Helpers/DialogHelper.cs:41]: done. Just a curiosity, are there any plans to migrate to gtk3?

File: Xamarin.Forms.Platform.UAP/Platform.cs [hunk 21]
```diff
@@ -579,7 +604,7 @@ static void OnPageActionSheet(object sender, ActionSheetArguments options)
 
 			try
 			{
-				actionSheet.ShowAt(((Page)sender).GetOrCreateRenderer().ContainerElement);
+				actionSheet.ShowAt(sender.GetOrCreateRenderer().ContainerElement);
```
Comment (VladislavAntonyuk) [inline Xamarin.Forms.Platform.UAP/Platform.cs:607]: sender is page, no need to cast

**Comment by VladislavAntonyuk** [comment]
> Nice work here @VladislavAntonyuk. I added some comments.
> We can split the PR in two ones, if you want. On the other hand, although it would not be necessary in this PR (we can add it later), it would be great to have some basic customization options (Example: `SetBackgroundColor`, `SetTextColor` and `SetActionTextColor`).

What is the best place to put this methods? I do not want to pass this settings through SnackbarArguments

**Comment by VladislavAntonyuk** [comment]
@samhouts, @jfversluis any changes required from my side?

**Comment by samhouts** [comment]
Since we're focusing exclusively on reliability and performance, we are unable to accept this contribution into Xamarin.Forms core. However, we would love to see this in the [Xamarin Community Toolkit](https://github.com/xamarin/XamarinCommunityToolkit). Would you be interested in moving this contribution there? Thank you so much for this effort--we'll do our best to make sure that it wasn't wasted!

**Comment by VladislavAntonyuk** [comment]
@samhouts I created a feature request and move my changes. But just a question will you move other alerts and prompts to XamarinComminityToolkit? Snackbar just a way to show message similar to alert so they should exist close to each other. 
Also what are the plans for XamarinComminityToolkit after MAUI release?

**Comment by samhouts** [comment]
@VladislavAntonyuk All good questions! We're still considering what should and should not move to the toolkit. I don't have a definite answer today. 

The Xamarin Community Toolkit should continue to exist long after .NET MAUI is released. We hope that most if not everything there will also work with .NET MAUI in the future.

Thanks!!

**Comment by DenisHidden** [comment]
hello how can i use snackbar now? still not released?

**Comment by VladislavAntonyuk** [comment]
@DenisHidden I moved code to XamarinCommunityToolkit repo. It is waiting for the code review

**Comment by jfversluis** [comment]
This is merged in the [XamarinCommunityToolkit](https://github.com/xamarin/XamarinCommunityToolkit) and we're shaping up the rest of the functionality and API, so please pitch in there! Thanks

**Comment by gitKRM** [comment]
don't  subscribe, I'm getting over 100 emails a day from this subscription
